CIPs/CIP-122.md

```diff
@@ -0,0 +1,454 @@
+```
+
+#### Updating the 3ID stream
+
+In order to make an update to the 3ID stream, one of the verification methods could delegate the following permission to a session key:
+
+```json
+{
+   "att":{
+      "did:3:z9CEZ1N1gbFK8J3rxVw3o6M5wygjoNFRSaEtkoGZw5fmbj": {
+        "did/vm-add": [],
+        "did/vm-remove": []
+      }
+			// CID of the event to append to the 3ID stream
+   },
+   "prf": [
+       { "/": "bafybeigk7ly3pog6uupxku3b6bubirr434ib6tfaymvox6gotaaaaaaaaa" }
+       // The CID of the capability that grants the key control over the 3ID
+       // From this the 3ID can be found, resolve revreg to see if self cap is revoked
+   ]
+}
+```
+
+#### Using a 3ID in Ceramic
+
+In order to act on behalf of a 3ID that controls a Ceramic stream, a delegation can be created by one of the verification methods in the 3ID. Note that the `prf` field must contain a link to the capability for the verification method.
+
+```json
+{
+   "tar":{
+      "ceramic://*?model=zkfif...": {
+        "crud/mutate": [],
+        "crud/read": []
+      }
+   },
+   "prf": [
+       { "/": "bafybeigk7ly3pog6uupxku3b6bubirr434ib6tfaymvox6gotaaaaaaaaa" }
+       // The CID of the capability that grants the key control over the 3ID
+   ]
+}
+```
+
+## Rationale
+
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+While most DID methods include full public keys, or other DIDs as verification methods 3ID chooses to use hashes of object-capabilities as verification methods. One of the main use cases for 3ID is to connect multiple PKH DIDs into a single identifier on Ceramic. While doing so publically is fine for some use cases, the ability to do so privately is quite important for many. Using *capability hashes* enables more privacy since the DIDs that are delegated to doesn't strictly need to be revealed. It is worth noting that revealing the CACAO object when used is the simplest way to prove a capability chain. However, it is possible to create zero-knowledge proofs that only reveal the hash of the capability used and which session key was delegated to.
+
+### Cryptographic Agility
+
+The 3 DID method itself doesn't really limit what cryptography can be used. It boils down to what the system that interprets the actual object capabilities is capable of. In Ceramic this includes the following, but is not limited to, and can be extended in the future:
+
+* `ed25519`
+* `secp256k1`
+* `secp256r1`
+
+Once good post quantum cryptography becomes more widely available extending Ceramic to support that will also be fairly straight forward.
+
+
+## Backwards Compatibility
```

Comment:
This is out of scope for this specification. This would be added on top, e.g. using a mechanism similar to `caip10-links`. See *Grafting DIDs* paragraph of this article: https://blog.ceramic.network/accounts-evolution-of-3id/

Reply:
I don't really follow how that helps with backwards compatibility?  Can you now use your 3id to update your streams that are controlled by your old did:pkh?

Reply:
> Can you now use your 3id to update your streams that are controlled by your old did:pkh?

Haven't thought about this use case. Interesting, but would need to think about how this could be done securely. Perhaps the `did:pkh` could delegate `*/* (or `did/control`) as a capability to the `did:3`.